stratumv2-lib/src/common/setup_connection.rs

```
@@ -0,0 +1,127 @@
+
+impl Serializable for Protocol {
+    fn serialize<W: io::Write>(&self, writer: &mut W) -> Result<usize> {
+        let length = u8::from(self).serialize(writer)?;
+
+        Ok(length)
+    }
+}
+
+impl Deserializable for Protocol {
+    fn deserialize(parser: &mut ByteParser) -> Result<Protocol> {
+        Protocol::try_from(u8::deserialize(parser)?)
+    }
+}
+
+pub enum SetupConnection {
+    Mining(mining::SetupConnection),
+    JobNegotiation(job_negotiation::SetupConnection),
+    // TemplateDistribution(template_distribution::SetupConnection),
+    // JobDistribution(job_distribution::SetupConnection),
+}
+
+impl Serializable for SetupConnection {
+    fn serialize<W: io::Write>(&self, writer: &mut W) -> Result<usize> {
+        let length = match self {
+            SetupConnection::Mining(v) => {
+                Protocol::Mining.serialize(writer)? + v.serialize(writer)?
+            }
+            SetupConnection::JobNegotiation(v) => {
+                Protocol::JobNegotiation.serialize(writer)? + v.serialize(writer)?
+            } // SetupConnection::TemplateDistribution(v) => {
+              //     Protocol::TemplateDistribution.serialize(writer)? + v.serialize(writer)?
+              // }
+              // SetupConnection::JobDistribution(v) => {
+              //     Protocol::JobDistribution.serialize(writer)? + v.serialize(writer)?
+              // }
+        };
+
+        Ok(length)
+    }
+}
+
+impl Deserializable for SetupConnection {
+    fn deserialize(parser: &mut ByteParser) -> Result<SetupConnection> {
+        let protocol = Protocol::deserialize(parser)?;
+        let variant = match protocol {
+            Protocol::Mining => {
+                SetupConnection::Mining(mining::SetupConnection::deserialize(parser)?)
+            }
+            Protocol::JobNegotiation => SetupConnection::JobNegotiation(
+                job_negotiation::SetupConnection::deserialize(parser)?,
+            ),
+            _ => panic!("Unimplemented"),
```

I think in this case, its better to return an Error enum, panic! will cause the program to exit

---

Done.